lib/WebService/Async/Segment.pm

```diff
@@ -0,0 +1,282 @@
+
+    return $self->{ua};
+}
+
+=head2 basic_authentication
+
+Settings required for basic HTTP authentication
+
+=cut
+
+sub basic_authentication {
+    my $self = shift;
+
+    #C<Net::Async::Http> basic authentication information
+    return {
+        user => $self->write_key // '',
+        pass => ''
+    };
+}
+
+=head2 method_call
+
+Makes a Segment method call. It automatically defaults C<sent_at> to the current time and C<< context->{library} >> to the current module.
+
+It takes the following named parameters:
+
+=over 4
+
+=item * C<method> - required. Segment method name (such as B<identify> and B<track>).
+
+=item * C<args> - optional. Method arguments represented as a dictionary. This may include either common, method-specific or custom fields.
+
+=back
+
+Please refer to L<https://segment.com/docs/spec/common/> for a full list of common fieds supported by Segment.
+
+It returns a L<Future> object that should be taken care of by the caller.
+
+=cut
+
+sub method_call {
+    my ($self, $method, %args) = @_;
+
+    $args{sent_at} = delete($args{sent_at}) // Time::Moment->now_utc->to_string();
+    $args{context}->{library}->{name}    = ref $self;
+    $args{context}->{library}->{version} = $VERSION;
+
+    return Future->fail('ValidationError', 'segment', 'Method name is missing', 'segment', $method, %args) unless $method;
+
+    return Future->fail('ValidationError', 'segment', 'Both user_id and anonymous_id are missing', $method, %args)
+        unless $args{user_id} or $args{anonymous_id};
+
+    %args = _snake_case_to_camelCase(\%args, SNAKE_FIELDS)->%*;
+
+    $log->tracef('Segment method %s called with params %s', $method, \%args);
+
+    return $self->ua->POST(
+        URI->new_abs($method, $self->base_uri),
+        encode_json_utf8(\%args),
+        content_type => 'application/json',
+        %{$self->basic_authentication},
+        )->then(
+        sub {
+            my $result = shift;
+
+            $log->tracef('Segment response for %s method received: %s', $method, $result);
+
+            my $response_str = $result->content;
+            return Future->fail('RequestFailed', 'segment', $response_str) unless $response_str =~ /^{.*}$/;
+
+            my $response = decode_json_utf8($response_str);
+            if ($response->{success}) {
+                $log->tracef('Segment %s method call finished successfully.', $method);
+
+                return Future->done($response->{success});
+            }
+            return Future->fail('RequestFailed', 'segment', $response_str);
+        }
+        )->on_fail(
+        sub {
+            $log->errorf('Segment method %s call failed: %s', $method, \@_);
+        });
+}
+
+=head2 new_customer
+
+Creates a new C<WebService::Async::Segment::Customer> object as the starting point of making B<identify> and B<track> calls.
+It may takes the following named standard arguments to populate the customer onject with:
+
+=over 4
+
+=item * C<userId> - Unique identifier of a user.
```

Comment:
These fields are inconsistent with the previous method? If we're going to use `snake_case` in some places, it should be consistently used *everywhere*.

Reply:
The inconsistency is in documentation not implementation. It will be fixed.